tensorflow/c/experimental/filesystem/plugins/gcs/ram_file_block_cache_test.cc

```diff
@@ -0,0 +1,589 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/c/experimental/filesystem/plugins/gcs/ram_file_block_cache.h"
+
+#include <cstring>
+
+#include "tensorflow/c/tf_status.h"
+#include "tensorflow/c/tf_status_internal.h"
+#include "tensorflow/core/lib/core/status_test_util.h"
+#include "tensorflow/core/platform/blocking_counter.h"
+#include "tensorflow/core/platform/cloud/now_seconds_env.h"
+#include "tensorflow/core/platform/notification.h"
+#include "tensorflow/core/platform/test.h"
+
+namespace tensorflow {
+namespace {
+
+Status ReadCache(tf_gcs_filesystem::RamFileBlockCache* cache, const string& filename,
+                 size_t offset, size_t n, std::vector<char>* out) {
+  out->clear();
+  out->resize(n, 0);
+  size_t bytes_transferred = 0;
+  TF_Status status;
+  cache->Read(filename, offset, n, out->data(), &bytes_transferred, &status);
+  EXPECT_LE(bytes_transferred, n);
+  out->resize(bytes_transferred, n);
+  return status.status;
+}
+
+TEST(RamFileBlockCacheTest, IsCacheEnabled) {
+  auto fetcher = [](const string& filename, size_t offset, size_t n,
+                    char* buffer, size_t* bytes_transferred,
+                    TF_Status* status) {
+    // Do nothing.
+    return TF_SetStatus(status, TF_OK, "");
+  };
+  tf_gcs_filesystem::RamFileBlockCache cache1(0, 0, 0, fetcher);
+  tf_gcs_filesystem::RamFileBlockCache cache2(16, 0, 0, fetcher);
+  tf_gcs_filesystem::RamFileBlockCache cache3(0, 32, 0, fetcher);
+  tf_gcs_filesystem::RamFileBlockCache cache4(16, 32, 0, fetcher);
+
+  EXPECT_FALSE(cache1.IsCacheEnabled());
+  EXPECT_FALSE(cache2.IsCacheEnabled());
+  EXPECT_FALSE(cache3.IsCacheEnabled());
+  EXPECT_TRUE(cache4.IsCacheEnabled());
+}
+
+TEST(RamFileBlockCacheTest, ValidateAndUpdateFileSignature) {
+  int calls = 0;
+  auto fetcher = [&calls](const string& filename, size_t offset, size_t n,
+                          char* buffer, size_t* bytes_transferred,
+                          TF_Status* status) {
+    calls++;
+    memset(buffer, 'x', n);
```

Comment:
Why not `memset` with 0? Or why `memset` at all? Shouldn't the caller be responsible for initial data here?

Reply:
No. The caller is responsible for allocating memory for `buffer`. But this function is a callback to fetch the data if the data we need isn't cached yet. So this function has to fill some data to `buffer`

Reply:
Here: https://github.com/tensorflow/tensorflow/blob/a61a816f86bb35d1d3a58f1c9a236be5e14c55c7/tensorflow/c/experimental/filesystem/plugins/gcs/ram_file_block_cache.h#L98-L99

And here:
https://github.com/tensorflow/tensorflow/blob/a61a816f86bb35d1d3a58f1c9a236be5e14c55c7/tensorflow/c/experimental/filesystem/plugins/gcs/ram_file_block_cache.h#L42-L50

Reply:
Oh, so we are mocking filling in the data. Sounds good.

Reply:
That's right